### PR TITLE
Support floating-point numbers without a fractional part

### PR DESCRIPTION
**Pull request by bw-hro**
Hello @benfrancis thanks for your fast response. Here you find a first approach of improving issue #62.

It is surprisingly hard to get the type information at runtime without changing the interface of existing classes. This is as Javas type erasure mechanism discards all type information of generics at compile time.

From what I can see, there are  two possible paths to go:

1. Workaround based on available meta information from the thing description:
  - For this path you can see a change suggestion in Thing.java
  - pro: does not break the interface
  - con: only supports integer input for doubles, floats will not work. Btw at the moment floats are also not working, even when fractional input is given. JSONObject always parses the fractional input to Double, which can't be cast to Float. But is Float really required, or will it be ok to officially only support Double?. This approach only works when "number" is given as type in thing description. It does not perform real type detection but just assumes that Double will be correct type.
   
2. Changing  some basic structures to make type information available at runtime: 
  - This could solve the issue in general and allows more smart helpers
  - To make type information available at runtime, we need to pass the Class<T> to Value<T> and make it available as member variable. This variable then could be used by Property and Thing. Unfortunately null is allowed as default value for Value class. So it is not save to query the class from default value at Value construction. A possible solution would be adding Class parameter to constructor or creating some custom annotation, but this breaks the interface and/or makes it verbose e.g.

```Java
Value<Double> val = new Value<>(Double.class, 12.3, theValue -> {}); // type passed in c'tor
Value<Double> val = new Value<>(Double.class, null, theValue -> {}); // type passed in c'tor
```

```Java
Value<Double> val = new Value<>(12.3, theValue -> {}); // type known from default value

@RememberType(Double.class) // create custom annotation
Value<Double> val = new Value<>(null, theValue -> {}); // annotation presence could be checked in c'tor in case default value is null
```

In addition I added an unit test to demonstrate the issue. (Float will still fail at the moment). So far my thoughts for the moment any comment welcome. Maybe I'm missing something and there is a simpler solution possible?

Best Regards 

Benno

### CODE REVIEW

**Comment by benfrancis** [comment]
Thanks for the PR!

This looks fine to me, and I think it's totally fine to rely on the metadata in the Thing Description, that's what it's there for.

I'm going to see if I can get a second review on this before landing though, because my Java is a bit rusty.

I think it's probably fine to always set a `number` from JSON to a `Double` in Java, but would like a second opinion. There may be some edge cases that are worth catching exceptions for.

**Comment by bw-hro** [comment]
Hi, I think I come up with a more general solution, that only  requires a miminal change in behavior. I hope to find some time today evening to provide a sample.

Small preview: 

When we accept, that `new Value<Double>(null, dv -> {})` is forbidden, then base type can be queried from within contructor. In case a client really want null as default value, he then needs to call the Value constructor like this: `new Value<>(Double.class, dv -> {})`  With some helpers, old null approch will fail with an explicit error message and requires minimal adjustment. Sample code still would work without any changes.

**Comment by bw-hro** [comment]
Hi, I created a more advanced solution. With this changes support of Long, Float in addition to Integer and Double is given. In junit test the api usage is demonstrated / tested. When a client used "null" as initial value for Value class in existing code, it wouldn't compile anymore, and the user would be forced to provide the Class instead of null. Everything else should work as before.

Let me know your opinion. 

Best regards, Benno

**Comment by bw-hro** [comment]
Hi, one more remark. Found out that the checks include building the project with Java 8 to 15. I think it would be more appropriate to check against Java 8, 11, 17, 21 as this are the long term support versions. What do you think?

also cf. https://www.oracle.com/java/technologies/java-se-support-roadmap.html

**Review by tim-hellhake — APPROVED** [review]
Looks good to me

**Comment by tim-hellhake** [comment]
> Hi, one more remark. Found out that the checks include building the project with Java 8 to 15. I think it would be more appropriate to check against Java 8, 11, 17, 21 as this are the long term support versions. What do you think?
> 
> also cf. https://www.oracle.com/java/technologies/java-se-support-roadmap.html

Makes sense, I would be happy the review another PR :)

**Comment by benfrancis** [comment]
Thank you @tim-hellhake !

**Pull request merged**
